### PR TITLE
Add support for custom endpoint

### PR DESCRIPTION
**Pull request by djberg96**
In order to support the Azure Stack environment we need to allow the users to set a custom endpoint since the default environments will not work. With the help of automatic discovery baked into the azure-armrest gem, we can support this fairly easily. If no endpoint is set then it simply defaults to finding the environment based on region as before.

Partially addresses https://bugzilla.redhat.com/show_bug.cgi?id=1594381

We will still need a modification on the UI side as well.

~~WIP for now until I add some specs.~~

### CODE REVIEW

File: spec/models/manageiq/providers/azure/cloud_manager_spec.rb [hunk 1]
```diff
@@ -76,6 +76,53 @@
...
+        @e.provider_region = "westus2"
+        @e.connect
+      end
+
+      it "with endpoint" do
+        expect(described_class).to receive(:raw_connect) do |clientid, clientkey, azure_tenant_id, subscription, http_proxy_uri, provider_region, endpoint|
+          expect(clientid).to eq("klmnopqrst")
+          expect(clientkey).to eq("1234567890")
+          expect(azure_tenant_id).to eq("abcdefghij")
+          expect(subscription).to eq("fghij67890")
```
Comment (bronaghs) [inline spec/models/manageiq/providers/azure/cloud_manager_spec.rb]: @djberg96 why are lines 111  through 114  repeated so many times in this spec?

File: spec/models/manageiq/providers/azure/cloud_manager_spec.rb [hunk 2]
```diff
@@ -76,6 +76,53 @@
...
+          expect(http_proxy_uri).to eq(nil)
+          expect(provider_region).to eq(nil)
+          expect(endpoint.url).to eq('http://www.foo.bar')
+          expect(endpoint.path).to eq('/')
+        end
+
+        @e.subscription = "fghij67890"
+        @e.endpoint = 'http://www.foo.bar'
+        @e.connect
+      end
```
Comment (bronaghs) [inline spec/models/manageiq/providers/azure/cloud_manager_spec.rb]: @djberg96 - how about a test for a valid endpoint *with* a path?

File: spec/models/manageiq/providers/azure/cloud_manager_spec.rb (hunk 1 above)
Comment (djberg96) [reply spec/models/manageiq/providers/azure/cloud_manager_spec.rb]: @bronaghs Just to make sure they weren't negatively affected. I can of course remove them if you feel they're redundant.

File: spec/models/manageiq/providers/azure/cloud_manager_spec.rb (hunk 2 above)
Comment (djberg96) [reply spec/models/manageiq/providers/azure/cloud_manager_spec.rb]: Ok, will add that.

**Comment by djberg96** [comment]
@bronaghs Ok, updated the specs, and even added a few more.

**Comment by djberg96** [comment]
@bronaghs @Ladas I actually want to undo some things first. It turns out there's already a `default_endpoint` method that will work for us, so this PR is a bit overworked atm.

File: app/models/manageiq/providers/azure/manager_mixin.rb [hunk 3]
```diff
@@ -25,6 +26,18 @@ def raw_connect(client_id, client_key, azure_tenant_id, subscription, proxy_uri
         $azure_log.warn("No region selected. Validating credentials against public environment.")
       end
 
+      endpoint_url = endpoint.respond_to?(:url) ? endpoint.url : endpoint.to_s
```
Comment (bronaghs) [inline app/models/manageiq/providers/azure/manager_mixin.rb]: `endpoint.try(:url) || endpoint.to_s` - maybe a little simpler

File: app/models/manageiq/providers/azure/manager_mixin.rb [hunk 4]
```diff
@@ -25,6 +26,18 @@ def raw_connect(client_id, client_key, azure_tenant_id, subscription, proxy_uri
...
+
+      if endpoint_url.present?
+        begin
+          environment = ::Azure::Armrest::Environment.discover(:url => endpoint_url, :proxy => proxy_uri)
+        rescue SocketError
+          raise MiqException::MiqInvalidCredentialsError, _("Invalid endpoint")
+        end
```
Comment (bronaghs) [inline app/models/manageiq/providers/azure/manager_mixin.rb]: Is it always a credentials error?

Comment (djberg96) [reply app/models/manageiq/providers/azure/manager_mixin.rb]: It basically means the endpoint is invalid, which I'm counting as "credentials" for our purposes. I actually couldn't find a better error, but I'm open to suggestions.

File: app/models/manageiq/providers/azure/manager_mixin.rb (hunk 3 above)
Comment (djberg96) [reply app/models/manageiq/providers/azure/manager_mixin.rb]: True, though I'm hoping https://github.com/ManageIQ/manageiq/pull/17678 will get merged so I can just call `to_s`.

File: app/models/manageiq/providers/azure/manager_mixin.rb (hunk 4 above)
Comment (djberg96) [reply app/models/manageiq/providers/azure/manager_mixin.rb]: Maybe this should be switched to an `MiqUnreachableError`.

Comment (bronaghs) [reply app/models/manageiq/providers/azure/manager_mixin.rb]: Agreed.

File: app/models/manageiq/providers/azure/manager_mixin.rb (hunk 3 above)
Comment (bronaghs) [reply app/models/manageiq/providers/azure/manager_mixin.rb]: Ok, but if ManageIQ/manageiq#17678  doesn't end up getting merged, can you simplify this line?

**Pull request merged**